src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs

```diff
@@ -0,0 +1,104 @@
+            };
+
+            cmd.HelpOption("-h|--help");
+
+            var argRoot = cmd.Argument(
+                    "[root]",
+                    "Optional path to a project file or MSBuild arguments.",
+                    multipleValues: true);
+
+            var sourceOption = cmd.Option(
+                    "-s|--source <source>",
+                    "Specifies a NuGet package source to use during the restore.",
+                    CommandOptionType.MultipleValue);
+
+            var packagesOption = cmd.Option(
+                    "--packages <packagesDirectory>",
+                    "Directory to install packages in.",
+                    CommandOptionType.SingleValue);
+
+            var disableParallelOption = cmd.Option(
+                    "--disable-parallel",
+                    "Disables restoring multiple projects in parallel.",
+                    CommandOptionType.NoValue);
+
+            var configFileOption = cmd.Option(
+                    "--configfile <file>",
+                    "The NuGet configuration file to use.",
+                    CommandOptionType.SingleValue);
+
+            var noCacheOption = cmd.Option(
+                    "--no-cache",
+                    "Do not cache packages and http requests.",
+                    CommandOptionType.NoValue);
+
+            var ignoreFailedSourcesOption = cmd.Option(
+                    "--ignore-failed-sources",
+                    "Treat package source failures as warnings.",
+                    CommandOptionType.NoValue);
+
+            cmd.OnExecute(() =>
+            {
+                var msbuildArgs = new List<string>()
+                {
+                     "/t:Restore"
+                };
+
+                if (sourceOption.HasValue())
+                {
+                    msbuildArgs.Add($"/p:RestoreSources={string.Join(";", sourceOption.Values)}");
+                }
+
+                if (packagesOption.HasValue())
+                {
+                    msbuildArgs.Add($"/p:RestorePackagesPath={packagesOption.Value()}");
+                }
+
+                if (disableParallelOption.HasValue())
+                {
+                    msbuildArgs.Add($"/p:RestoreDisableParallel=true");
+                }
+
+                if (configFileOption.HasValue())
+                {
+                    msbuildArgs.Add($"/p:RestoreConfigFile={configFileOption.Value()}");
+                }
+
+                if (noCacheOption.HasValue())
+                {
+                    msbuildArgs.Add($"/p:RestoreNoCache=true");
+                }
+
+                if (ignoreFailedSourcesOption.HasValue())
+                {
+                    msbuildArgs.Add($"/p:RestoreIgnoreFailedSources=true");
+                }
+
+                // Add any additional values passed
+                msbuildArgs.AddRange(argRoot.Values);
+
+                return new MSBuildForwardingApp(msbuildArgs.ToArray()).Execute();
```

Comment:
There's no need that MSBuildForwardingApp strictly needs a `string[]`.  We should change that class to take an IEnumerable<string> instead, and then get rid of the ToArray() call here.